File: hibernate-query-language/src/main/java/io/github/robertomike/hefesto/constructors/ConstructWhereImplementation.java
package io.github.robertomike.hefesto.constructors;

import io.github.robertomike.hefesto.actions.wheres.*;
import io.github.robertomike.hefesto.builders.Hefesto;
import io.github.robertomike.hefesto.exceptions.QueryException;
import io.github.robertomike.hefesto.exceptions.UnsupportedOperationException;

import java.util.*;

public class ConstructWhereImplementation extends ConstructWhere {
    private Map<String, Object> params;

    public String construct(Map<String, Object> params) {
        this.params = params;

        if (items.isEmpty()) {
            return "";
        }
        List<String> wheresQuery = new ArrayList<>();

        iterator(wheresQuery, items);

        if (!wheresQuery.isEmpty()) {
            wheresQuery.add(0, "Where ");
        }

        return String.join("", wheresQuery);
    }


    public void iterator(List<String> wheresQuery, List<? extends BaseWhere> whereList) {
        whereList.forEach((value) -> {
            if (value instanceof CollectionWhere) {
                if (!wheresQuery.isEmpty()) {
                    wheresQuery.add(" " + value.getWhereOperation().operator);
                }

                wheresQuery.add(" (");
                iterator(wheresQuery, ((CollectionWhere) value).getWheres());
                wheresQuery.add(")");
                return;
            }
            constructBaseWhere(wheresQuery, value);
        });
    }

    public void constructBaseWhere(List<String> wheresQuery, BaseWhere where) {
        if (!wheresQuery.isEmpty()) {
            if (!wheresQuery.get(wheresQuery.size() - 1).equals(" ("))
                wheresQuery.add(" " + where.getWhereOperation().operator);
        }

        wheresQuery.add(" ");

        if (where instanceof WhereRaw whereRaw) {
            wheresQuery.add(whereRaw.query);
            return;
        }

        if (where instanceof WhereExist whereExist) {
            wheresQuery.add(whereExist.isExists() ? "exists" : "not exists");
            wheresQuery.add("(");
            wheresQuery.add(((Hefesto<?>) whereExist.getSubQuery()).getSubQuery(params));
            wheresQuery.add(")");
            return;
        }

        if (where instanceof WhereField whereField) {
            constructWhereField(wheresQuery, whereField);
            return;
        }

        if (where instanceof Where nativeWhere) {
            constructWhere(wheresQuery, nativeWhere);
            return;
        }

        throw new QueryException("Invalid class extended from BaseWhere: " + where.getClass());
    }

    private void constructWhereField(List<String> wheresQuery, WhereField whereField) {
        switch (whereField.getOperator()) {
            case LIKE, NOT_LIKE, EQUAL, DIFF, GREATER_OR_EQUAL, LESS, LESS_OR_EQUAL ->
                    wheresQuery.add(whereField.getParentField() + " " + whereField.getOperator().operator + " " + whereField.getField());

            default -> throw new UnsupportedOperationException("Unsupported operator: " + whereField.getOperator());
        }
    }

    public String standardizeNameParam(String field, int size) {
        return field.replace(".", "")
                .replace("(", "_")
                .replace(")", "_") + "_" + size;
    }

    public String standardizeNameParamWhere(String nameParam) {
        return ":" + nameParam;
    }

    public void constructWhere(List<String> wheresQuery, Where where) {
        String field = where.getField();
        String nameParam = standardizeNameParam(field, wheresQuery.size());
        String nameParamWhere = standardizeNameParamWhere(nameParam);
        String operator = where.getOperator().operator;
        var value = where.getValue();

        boolean param = true;

        switch (where.getOperator()) {
            case IN, NOT_IN -> {
                if (where.getValue() instanceof Hefesto<?> builder) {
                    wheresQuery.add(field + " " + operator + " (" + builder.getSubQuery(params) + ")");
                    return;
                }

                wheresQuery.add(field + " " + operator + " (" + nameParamWhere + ")");

                Object values;
                if (value.getClass().isArray()) {
                    values = Arrays.asList((Object[]) value);
                } else if (value instanceof Collection<?> collection) {
                    values = new ArrayList<>(collection);
                } else if (where.getValue() instanceof Hefesto<?> builder) {
                    values = builder.getSubQuery(params);
                } else {
                    throw new UnsupportedOperationException("Invalid class: " + value.getClass());
                }

                params.put(nameParam, values);
                param = false;
            }

            case IS_NULL, IS_NOT_NULL -> {
                wheresQuery.add(field + " " + operator);
                param = false;
            }

            case NOT_FIND_IN_SET ->
                wheresQuery.add(operator + "(" + nameParamWhere + "," + field + ") = 0");
            case FIND_IN_SET ->
                wheresQuery.add(operator + "(" + nameParamWhere + "," + field + ") > 0");

            default -> wheresQuery.add(field + " " + operator + " " + nameParamWhere);
        }

        if (param) {
            params.put(nameParam, value);
        }
    }
}

File: hibernate-query-language/src/main/kotlin/io/github/robertomike/hefesto/constructors/ConstructWhereImplementation.kt
package io.github.robertomike.hefesto.constructors

import io.github.robertomike.hefesto.actions.wheres.*
import io.github.robertomike.hefesto.builders.Hefesto
import io.github.robertomike.hefesto.exceptions.QueryException
import io.github.robertomike.hefesto.exceptions.UnsupportedOperationException

class ConstructWhereImplementation : ConstructWhere() {
    private lateinit var params: MutableMap<String, Any?>

    fun construct(params: MutableMap<String, Any?>): String {
        this.params = params

        if (items.isEmpty()) {
            return ""
        }
        val wheresQuery = mutableListOf<String>()

        iterator(wheresQuery, items)

        if (wheresQuery.isNotEmpty()) {
            wheresQuery.add(0, "Where ")
        }

        return wheresQuery.joinToString("")
    }

    fun iterator(wheresQuery: MutableList<String>, whereList: List<BaseWhere>) {
        whereList.forEach { value ->
            if (value is CollectionWhere) {
                if (wheresQuery.isNotEmpty()) {
                    wheresQuery.add(" ${value.whereOperation.operator}")
                }

                wheresQuery.add(" (")
                iterator(wheresQuery, value.wheres)
                wheresQuery.add(")")
                return@forEach
            }
            constructBaseWhere(wheresQuery, value)
        }
    }

    fun constructBaseWhere(wheresQuery: MutableList<String>, where: BaseWhere) {
        if (wheresQuery.isNotEmpty()) {
            if (wheresQuery[wheresQuery.size - 1] != " (") {
                wheresQuery.add(" ${where.whereOperation.operator}")
            }
        }

        wheresQuery.add(" ")

        if (where is WhereRaw) {
            wheresQuery.add(where.query)
            return
        }

        if (where is WhereExist) {
            wheresQuery.add(if (where.exists) "exists" else "not exists")
            wheresQuery.add("(")
            wheresQuery.add((where.subQuery as Hefesto<*>).getSubQuery(params))
            wheresQuery.add(")")
            return
        }

        if (where is WhereField) {
            constructWhereField(wheresQuery, where)
            return
        }

        if (where is Where) {
            constructWhere(wheresQuery, where)
            return
        }

        throw QueryException("Invalid class extended from BaseWhere: ${where.javaClass}")
    }

    private fun constructWhereField(wheresQuery: MutableList<String>, whereField: WhereField) {
        when (whereField.operator) {
            io.github.robertomike.hefesto.enums.Operator.LIKE,
            io.github.robertomike.hefesto.enums.Operator.NOT_LIKE,
            io.github.robertomike.hefesto.enums.Operator.EQUAL,
            io.github.robertomike.hefesto.enums.Operator.DIFF,
            io.github.robertomike.hefesto.enums.Operator.GREATER_OR_EQUAL,
            io.github.robertomike.hefesto.enums.Operator.LESS,
            io.github.robertomike.hefesto.enums.Operator.LESS_OR_EQUAL ->
                wheresQuery.add("${whereField.parentField} ${whereField.operator.operator} ${whereField.field}")

            else -> throw UnsupportedOperationException("Unsupported operator: ${whereField.operator}")
        }
    }

    fun standardizeNameParam(field: String, size: Int): String {
        return field.replace(".", "")
            .replace("(", "_")
            .replace(")", "_") + "_$size"
    }

    fun standardizeNameParamWhere(nameParam: String): String {
        return ":$nameParam"
    }

    fun constructWhere(wheresQuery: MutableList<String>, where: Where) {
        val field = where.field
        val nameParam = standardizeNameParam(field, wheresQuery.size)
        val nameParamWhere = standardizeNameParamWhere(nameParam)
        val operator = where.operator.operator
        val value = where.value

        var param = true

        when (where.operator) {
            io.github.robertomike.hefesto.enums.Operator.IN,
            io.github.robertomike.hefesto.enums.Operator.NOT_IN -> {
                if (where.value is Hefesto<*>) {
                    wheresQuery.add("$field $operator (${(where.value as Hefesto<*>).getSubQuery(params)})")
                    return
                }

                wheresQuery.add("$field $operator ($nameParamWhere)")

                val values: Any = when {
                    value?.javaClass?.isArray == true -> (value as Array<*>).toList()
                    value is Collection<*> -> ArrayList(value)
                    where.value is Hefesto<*> -> (where.value as Hefesto<*>).getSubQuery(params)
                    else -> throw UnsupportedOperationException("Invalid class: ${value?.javaClass}")
                }

                params[nameParam] = values
                param = false
            }

            io.github.robertomike.hefesto.enums.Operator.IS_NULL,
            io.github.robertomike.hefesto.enums.Operator.IS_NOT_NULL -> {
                wheresQuery.add("$field $operator")
                param = false
            }

            io.github.robertomike.hefesto.enums.Operator.NOT_FIND_IN_SET ->
                wheresQuery.add("$operator($nameParamWhere,$field) = 0")

            io.github.robertomike.hefesto.enums.Operator.FIND_IN_SET ->
                wheresQuery.add("$operator($nameParamWhere,$field) > 0")

            else -> wheresQuery.add("$field $operator $nameParamWhere")
        }

        if (param) {
            params[nameParam] = value
        }
    }
}
